### ISSUE TITLE
Javascript code generation with python.tools.protoc error

### ISSUE THREAD

**Issue by lromor** [grpc/grpc#8135]
Hello, 
I'm having this error while trying to generate a javascript output with the python grpcio-tool:
(python -m grpc.tools.protoc etc..)

```
 protoc-gen-js: program not found or is not executable
--js_out: protoc-gen-js: Plugin failed with status code 1.
```

using the normal protoc binary command  with the same arguments works:

```
protoc --proto_path=. --js_out=import_style=commonjs,binary:static/ --grpc_out=static/ --plugin=protoc-gen-grpc=/usr/local/bin/grpc_node_plugin interface.proto
```

Could you help me to understand what's going on? 
thanks,

-l

**Comment by nathanielmanistaatgoogle** [comment]
I don't think it's possible to use the Python package `grpcio-tools` to generate code for languages other than Python, but I think it's seriously cool that you tried. Is there some documentation somewhere that mislead you into thinking that `grpcio-tools` supported generating non-Python code? We'd be happy to fix that up.

Can you tell us more about your use case? In what situation are you finding yourself that has you wanting to do this? How painful for you is it that you can't?

**Comment by lromor** [comment]
Given than I just started developing code with grpc (so I'm still in a very early learning phase, so forgive me if I will say stupid things) and that there is no misleading documentation that made me test code generation for js, except the name of the tool itself and its arguments (I thought was some kind of a wrapper), what I'm trying to do is to write my python code so that one day I may be able to autogenerate a GUI given a certain proto (for the moment when grpc will be available even for web browsers ,my python code has plugins and I want to automatize the interface generation).

So in order to do that, I though (for now) about the following solutions:
a) I will serve the .proto from my server and let the browser dynamically load it (such as in the node environment) and uhm, well, generate the interface, or
b) I could generate the whole js/html code ready to be loaded from the browser. 

Another interesting thing for me would be to add cython-like types, so that instead
of writing a proto, compiling it, loading the *_pb.py, and register a servicer to the server, maybe with a bunch of decorators and declared types there could be an easier mapping between a class and a service (maybe I'm just delirious) reducing in this way the amount of redundant operations to do.

**Comment by soltanmm-google** [comment]
gRPC *might* be a confounder for doing what you want to do. Until we have gRPC in the web browser, it sounds like you're better off serving the proto file just like any other file and using some implementation of JavaScript protobuf to parse it and use it to generate your GUI if you're going with option A. Option B doesn't appear to have an intrinsic RPC-element to it.

The last bit about having decorator-wrappers around gRPC internals (like Flask) is a nice idea. I don't think it fits in with the main body of gRPC Python, but I can see that sort of super-convenient zero-to-sixty-in-oh-point-two seconds thing fitting one day into e.g. https://github.com/grpc/grpc-contrib.

**Comment by lromor** [comment]
> gRPC might be a confounder for doing what you want to do. Until we have gRPC in the web browser, it sounds like you're better off serving the proto file just like any other file and using some implementation of JavaScript protobuf to parse it and use it to generate your GUI if you're going with option A. Option B doesn't appear to have an intrinsic RPC-element to it.

Thanks, that gives me a direction to follow.

> The last bit about having decorator-wrappers around gRPC internals (like Flask) is a nice idea. I don't think it fits in with the main body of gRPC Python, but I can see that sort of super-convenient zero-to-sixty-in-oh-point-two seconds thing fitting one day into e.g. https://github.com/grpc/grpc-contrib.

ok, I see what you mean, well, thanks @nathanielmanistaatgoogle and @soltanmm-google

**Comment by lromor** [comment]
Hello, 
I would like to reopen this since I started writing a snippet for this decorator and new python annotations.

I also found out that in google-cloud a similar system was already provided (If I understood correctly) (https://cloud.google.com/endpoints/docs/frameworks/python/create_api)
with "protorpc"?

So I'm stuck now at a point where I can create programmatically a server but then It's not really "exportable" to any client without creating a matching .proto.

Ideally in the "\@Rpc" decorator there should be a facility that help you writing a compilable .proto. I guess there's no way to generate back a .proto right?
